File: security_cam/lib/event/trigger_batcher.dart
import 'dart:async';

import '../core/models.dart';

class TriggerBatch {
  final DateTime timestamp;
  final List<TriggerEvent> triggers;
  final Snapshot? snapshot;
  final String? videoName;

  TriggerBatch({
    required this.timestamp,
    required this.triggers,
    this.snapshot,
    this.videoName,
  });
}

class TriggerBatcher {
  final Duration window;
  final Future<Snapshot?> Function() captureSnapshot;

  /// Optional video clip capture started on the first trigger of a batch
  /// (Android only; null elsewhere). Receives the trigger timestamp so the
  /// native side can bound the pre-roll ring buffer, and resolves to the clip
  /// display name (or null) once the post-roll tail is recorded.
  final Future<String?> Function(DateTime triggerAt)? captureVideo;

  final StreamController<TriggerBatch> _batches =
      StreamController<TriggerBatch>.broadcast();
  DateTime? _openedAt;
  List<TriggerEvent> _pending = [];
  Future<Snapshot?>? _pendingSnapshot;
  Future<String?>? _pendingVideo;
  Timer? _timer;
  bool _disposed = false;

  TriggerBatcher({
    required this.window,
    required this.captureSnapshot,
    this.captureVideo,
  });

  Stream<TriggerBatch> get batches => _batches.stream;

  void add(TriggerEvent event) {
    if (_disposed) return;
    if (_pending.isEmpty) {
      _openedAt = event.timestamp;
      _pendingSnapshot = _captureSnapshotSafely();
      _pendingVideo = _captureVideoSafely(event.timestamp);
      _timer = Timer(window, () => unawaited(_flush()));
    }
    _pending.add(event);
  }

  Future<Snapshot?> _captureSnapshotSafely() async {
    try {
      return await captureSnapshot();
    } catch (_) {
      return null;
    }
  }

  Future<String?> _captureVideoSafely(DateTime triggerAt) async {
    final capture = captureVideo;
    if (capture == null) return null;
    try {
      return await capture(triggerAt);
    } catch (_) {
      return null;
    }
  }

  Future<void> _flush() async {
    _timer?.cancel();
    _timer = null;
    if (_pending.isEmpty) return;
    final openedAt = _openedAt!;
    final events = List<TriggerEvent>.unmodifiable(_pending);
    final snapshotFuture = _pendingSnapshot;
    final videoFuture = _pendingVideo;
    _pending = [];
    _pendingSnapshot = null;
    _pendingVideo = null;
    _openedAt = null;
    final snapshot = snapshotFuture == null ? null : await snapshotFuture;
    final videoName = videoFuture == null ? null : await videoFuture;
    _batches.add(TriggerBatch(
      timestamp: openedAt,
      triggers: events,
      snapshot: snapshot,
      videoName: videoName,
    ));
  }

  Future<void> dispose() async {
    _disposed = true;
    _timer?.cancel();
    _timer = null;
    await _batches.close();
  }
}

File: level2/android/app/src/main/kotlin/io/securitycam/level2/event/TriggerBatcher.kt
package io.securitycam.level2.event

import io.securitycam.level2.core.Snapshot
import io.securitycam.level2.core.TriggerEvent
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.Duration
import java.time.Instant

/** A merged batch of triggers plus one snapshot/clip, emitted per window. */
data class TriggerBatch(
    val timestamp: Instant,
    val triggers: List<TriggerEvent>,
    val snapshot: Snapshot? = null,
    val videoName: String? = null,
)

/**
 * Merges triggers within a window into one batch (port of
 * `lib/event/trigger_batcher.dart`). A snapshot capture starts on the first
 * trigger of a batch; an optional clip export starts too and resolves to its
 * display name once the post-roll tail is recorded.
 */
class TriggerBatcher(
    private val scope: CoroutineScope,
    private val window: Duration,
    private val captureSnapshot: suspend () -> Snapshot?,
    private val captureVideo: suspend (Instant) -> String? = { null },
) {
    private val batchFlow = MutableSharedFlow<TriggerBatch>(
        extraBufferCapacity = 16,
        onBufferOverflow = kotlinx.coroutines.channels.BufferOverflow.DROP_OLDEST,
    )

    val batches: Flow<TriggerBatch> get() = batchFlow.asSharedFlow()

    private val mutex = Mutex()
    private var openedAt: Instant? = null
    private val pending = ArrayList<TriggerEvent>()
    private var pendingSnapshot: Deferred<Snapshot?>? = null
    private var pendingVideo: Deferred<String?>? = null
    private var timer: Job? = null

    @Volatile
    private var disposed = false

    fun add(event: TriggerEvent) {
        if (disposed) return
        scope.launch {
            mutex.withLock {
                if (disposed) return@withLock
                if (pending.isEmpty()) {
                    openedAt = event.timestamp
                    pendingSnapshot = scope.async {
                        try {
                            captureSnapshot()
                        } catch (e: CancellationException) {
                            throw e
                        } catch (_: Exception) {
                            null
                        }
                    }
                    pendingVideo = scope.async {
                        try {
                            captureVideo(event.timestamp)
                        } catch (e: CancellationException) {
                            throw e
                        } catch (_: Exception) {
                            null
                        }
                    }
                    timer = scope.launch {
                        delay(window.toMillis())
                        flush()
                    }
                }
                pending.add(event)
            }
        }
    }

    private suspend fun flush() {
        val events: List<TriggerEvent>
        val snapshotFuture: Deferred<Snapshot?>?
        val videoFuture: Deferred<String?>?
        val batchOpenedAt: Instant
        mutex.withLock {
            // flush() runs INSIDE the timer coroutine, so cancelling `timer`
            // here would cancel this very coroutine: the next real suspension
            // point (awaiting a capture below) would throw
            // CancellationException and silently drop the batch. Just drop
            // the reference; dispose() cancels a still-pending timer.
            timer = null
            if (pending.isEmpty()) return
            batchOpenedAt = openedAt!!
            events = ArrayList(pending)
            snapshotFuture = pendingSnapshot
            videoFuture = pendingVideo
            pending.clear()
            pendingSnapshot = null
            pendingVideo = null
            openedAt = null
        }
        val snapshot = snapshotFuture?.await()
        val videoName = videoFuture?.await()
        // A concurrent dispose() may have run while awaiting; drop the batch.
        if (disposed) return
        batchFlow.tryEmit(TriggerBatch(batchOpenedAt, events, snapshot, videoName))
    }

    suspend fun dispose() {
        mutex.withLock {
            disposed = true
            timer?.cancel()
            timer = null
        }
    }
}